piecrust-uplink/src/lib.rs

```diff
@@ -4,13 +4,69 @@
 //
 // Copyright (c) DUSK NETWORK. All rights reserved.
 
+//! Piecrust Uplink is the crate for writing WASM smart-contracts in Rust
+//! targeting the Piecrust virtual machine.
+//!
+//! A smart-contract is a program exposing a collection of functions that
+//! operate on a shared state. These functions can perform a variety of
+//! operations, such as reading or modifying the state, or interacting with the
+//! host.
+//!
+//! # State Model and ABI
+//! Contracts targeting the Piecrust VM represent their state as a WASM memory.
```

Comment:
Could be valuable to link to an explainer on `WASM memory` like this one: https://wasmbyexample.dev/examples/webassembly-linear-memory/webassembly-linear-memory.rust.en-us.html